fluture.js

```diff
@@ -376,544 +404,927 @@
+  Future.prototype[FL.chain] = Future.prototype.chain;
+  Future.prototype[FL.chainRec] = Future.prototype.chainRec;
+
+  /////////////////
+  // Sub classes //
+  /////////////////
+
+  function SafeFuture(computation){
+    this._computation = computation;
   }
 
-  function Future$safeFork(rej, res){
+  SafeFuture.prototype = Object.create(Future.prototype);
+
+  SafeFuture.prototype._f = function SafeFuture$fork(rej, res){
     let open = true;
-    const f = this._raw(function Future$safeFork$rej(x){
+    const f = this._computation(function SafeFuture$fork$rej(x){
       if(open){
         open = false;
         rej(x);
       }
-    }, function Future$safeFork$res(x){
+    }, function SafeFuture$fork$res(x){
       if(open){
         open = false;
         res(x);
       }
     });
-    check$fork$f(f, this._raw);
-    return function Future$safeFork$cancel(){
+    check$fork$f(f, this._computation);
+    return function SafeFuture$fork$cancel(){
       open && f && f();
       open = false;
     };
   }
 
-  function Future$chain(f){
-    check$chain(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$chain$fork(rej, res){
-      let cancel;
-      const r = _this._f(rej, function Future$chain$res(x){
-        const m = f(x);
-        check$chain$f(m, f, x);
-        cancel = m._f(rej, res);
-      });
-      return cancel ? cancel : (cancel = r, function Future$chain$cancel(){ cancel() });
-    });
+  SafeFuture.prototype.toString = function SafeFuture$toString(){
+    return `Future(${showf(this._computation)})`;
   }
 
-  function Future$recur(f){
-    check$recur(this, f);
-    const _this = this;
-    return new FutureClass(false, function Future$chain$fork(rej, res){
-      return _this._f(rej, function Future$chain$res(x){
-        const m = f(x);
-        check$recur$f(m, f, x);
-        m._f(rej, res);
-      });
-    });
+  //----------
+
+  function ChainRec(iterate, init){
+    this._iterate = iterate;
+    this._init = init;
   }
 
-  function Future$chainRej(f){
-    check$chainRej(this, f);
+  ChainRec.prototype = Object.create(Future.prototype);
+
+  ChainRec.prototype._f = function ChainRec$fork(rej, res){
     const _this = this;
-    return new FutureClass(false, function Future$chainRej$fork(rej, res){
-      let cancel;
-      const r = _this._f(function Future$chainRej$rej(x){
-        const m = f(x);
-        check$chainRej$f(m, f, x);
-        cancel = m._f(rej, res);
-      }, res);
-      return cancel ? cancel : (cancel = r, function Future$chainRej$cancel(){ cancel() });
-    });
+    let cancel = noop, i = 0;
+    (function Future$chainRec$recur(state){
+      let isSync = null;
+      function Future$chainRec$res(it){
+        check$chainRec$it(it, i);
+        i = i + 1;
+        if(isSync === null){
+          isSync = true;
+          state = it;
+        }else{
+          Future$chainRec$recur(it);
+        }
+      }
+      while(!state.done){
+        isSync = null;
```

Comment:
Numbers would be an improvement. This would be even clearer:

```javascript
//    data Timing = Synchronous | Asynchronous | Undetermined

//    Synchronous :: Timing
const Synchronous = 'Synchronous';

//    Asynchronous :: Timing
const Asynchronous = 'Asynchronous';

//    Undetermined :: Timing
const Undetermined = 'Undetermined';
```